### PR TITLE
Fixed WOM offset when n=0

### LINKED ISSUE #20517: date_range fails when I try to generate ones with 1 periods and freq equal WOM-1MON
#### Code Sample

```python
import pandas as pd
pd.date_range('20100104', periods=2, freq='WOM-1MON')  # works
pd.date_range('20100104', periods=1, freq='WOM-1MON')  # fails
```

```python-traceback
Traceback (most recent call last):
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2910, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-31-ec0b4bad59c9>", line 1, in <module>
    pd.date_range('20100104', periods=1, freq='WOM-1MON')
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimes.py", line 2057, in date_range
    closed=closed, **kwargs)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/util/_decorators.py", line 118, in wrapper
    return func(*args, **kwargs)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimes.py", line 324, in __new__
    ambiguous=ambiguous)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimes.py", line 531, in _generate
    index = _generate_regular_range(start, end, periods, offset)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/core/indexes/datetimes.py", line 2009, in _generate_regular_range
    dates = list(xdr)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 2960, in generate_range
    end = start + (periods - 1) * offset
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 425, in __rmul__
    return self.__mul__(someInt)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 422, in __mul__
    **self.kwds)
  File "/Users/mmngreco/miniconda3/envs/pandas-dev/lib/python3.6/site-packages/pandas/tseries/offsets.py", line 1671, in __init__
    raise ValueError('N cannot be 0')
ValueError: N cannot be 0
```
#### Problem description

If N is equal to periods then is not 0 as we can see, that make me think that probably there is something wrong in the code.

#### Expected Output

```python
Out[33]: DatetimeIndex(['2010-01-04'], dtype='datetime64[ns]', freq='WOM-1MON')
```

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Darwin
OS-release: 17.4.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: es_ES.UTF-8
pandas: 0.22.0
pytest: 3.4.2
pip: 9.0.1
setuptools: 38.5.1
Cython: 0.27.3
numpy: 1.14.2
scipy: 1.0.0
pyarrow: 0.9.0
xarray: 0.10.2
IPython: 6.2.1
sphinx: 1.7.1
patsy: 0.5.0
dateutil: 2.7.0
pytz: 2018.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.2.2
openpyxl: 2.5.1
xlrd: 1.1.0
xlwt: 1.2.0
xlsxwriter: 1.0.2
lxml: 4.2.0
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.5
pymysql: 0.8.0
psycopg2: None
jinja2: 2.10
s3fs: 0.1.3
fastparquet: 0.1.4
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by asdf8601**
Checklist for other PRs (remove this part if you are doing a PR for the pandas documentation sprint):

- [x] closes #20517
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
@mmngreco can you add a whatsnew note (other enhancements section), and also a test for the ``date_range`` (both ones) as you indicated in the issue. otherwise lgtm. ping on green.

**Comment by mmngreco** [comment]
Hi @jreback, I see that travis fails in https://travis-ci.org/pandas-dev/pandas/jobs/360472961, any idea whats going wrong?

**Comment by mmngreco** [comment]
ping @jreback

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -403,8 +403,10 @@ Other Enhancements
   ``SQLAlchemy`` dialects supporting multivalue inserts include: ``mysql``, ``postgresql``, ``sqlite`` and any dialect with ``supports_multivalues_insert``. (:issue:`14315`, :issue:`8953`)
 - :func:`read_html` now accepts a ``displayed_only`` keyword argument to controls whether or not hidden elements are parsed (``True`` by default) (:issue:`20027`)
 - zip compression is supported via ``compression=zip`` in :func:`DataFrame.to_pickle`, :func:`Series.to_pickle`, :func:`DataFrame.to_csv`, :func:`Series.to_csv`, :func:`DataFrame.to_json`, :func:`Series.to_json`. (:issue:`17778`)
+- Now is possible create a :class:`WeekOfMonth` offset with `n=0` (:issue:`20517`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: re-word this something like: ``:class:`WeekOfMonth` `` constuctor will support ``n=0``

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 2]
```diff
@@ -236,6 +236,10 @@ def test_catch_infinite_loop(self):
         pytest.raises(Exception, date_range, datetime(2011, 11, 11),
                       datetime(2011, 11, 12), freq=offset)
 
+    def test_wom_len_one(self):
+        # https://github.com/pandas-dev/pandas/issues/20517
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: can you parametrize on periods=1, 2

**Review by jreback — CHANGES_REQUESTED** [review]
small comments, otherwise lgtm. ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mmngreco

**Pull request merged**